Products: Add duplicate product feature

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #7255 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds the feature to duplicate products.

This feature is available for all products that have been saved on the server, regardless of their status (published/draft).

This feature is implemented on WooCommerce Core [here](https://github.com/woocommerce/woocommerce/blob/5d7f6acbcb387f1d51d51305bf949d07fa3c4b08/includes/admin/class-wc-admin-duplicate-product.php#L97), but not available on the API.

The idea is simply to create a copy of the original product, rename it, set its status to draft, and make sure to clear the SKU value since it needs to be unique. The implementation on Core handles generating a new unique value based on the original one, but this requires checking for existing values which may take time, so I'm just resetting it to nil for simplicity.

For variable products, the original product's variations are also duplicated to the new one.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Navigate to the Products tab, and select a simple product with SKU value. 
- Tap the "..." button on the top right and select Duplicate. Notice in Xcode console: `🔵 Tracked product_detail_duplicate_button_tapped`
-  Notice that after the process completes, the view is reloaded with the copied product, with the name updated and no SKU value. The status of the product should be Draft. Most of the other values should be the same as the original one, except for reviews, rating, slug, and product ID. Xcode console should log: `🔵 Tracked duplicate_product_success`
- Repeat the test with a variable product. After duplicating, the new product should have the same variations as the original.
- Turn off the Internet connection and try duplicating product again. Xcode console should log `🔵 Tracked duplicate_product_failed` with error information in the properties.

### Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://user-images.githubusercontent.com/5533851/190574415-498989a5-e2b3-4e83-928e-94d8397c70d2.mp4



---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

itsmeichigo committed:
Add canDuplicateProduct to ProductFormViewModelProtocol

itsmeichigo committed:
Add duplicate product to action sheet

itsmeichigo committed:
Implement duplicate product

itsmeichigo committed:
Call duplicate product from the product form

itsmeichigo committed:
Show progress view when duplicating product

itsmeichigo committed:
Update status of copied product to draft

itsmeichigo committed:
Update ProductFormRemoteActionUseCase with duplicate product

itsmeichigo committed:
Use remoteActionUseCase duplicateProduct

itsmeichigo committed:
Add tests for canDuplicateProduct

itsmeichigo committed:
Also update password for duplicated product

itsmeichigo committed:
Fix issue duplicating product with sku

itsmeichigo committed:
Fix issue building on CI

itsmeichigo committed:
Update unit tests for ProductFormRemoteActionUseCase

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormRemoteActionUseCase.swift:
```diff
@@ -205,4 +248,89 @@ private extension ProductFormRemoteActionUseCase {
+                    continuation.resume(returning: nil)
+                }
+            }
+            DispatchQueue.main.async { [weak self] in
+                self?.stores.dispatch(action)
+            }
```
Actions need to be dispatched on the main thread, but I don't know how to do that with just async-await syntax, so I'm using `DispatchQueue` here. Please feel free to suggest a better solution if possible.

itsmeichigo committed:
Add tracks

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController+Helpers.swift:
```diff
@@ -35,14 +35,16 @@ extension ProductFormViewController {
     /// Product Confirmation Save alert
     ///
-    func presentProductConfirmationSaveAlert() {
+    func presentProductConfirmationSaveAlert(hasDuplicated: Bool) {
```
Super nits:
1. What do you think about naming the param as something like `wasProductDuplicated` for clarity?
2. Could we also update the comment to include a point about the parameter?

itsmeichigo replied:
I switched to sending the alert title as a parameter to make it simpler in 7b6b744bd4971385712864b60c212c3be21d941f.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController+Helpers.swift:
```diff
@@ -170,6 +179,11 @@ private enum Localization {
                                                             comment: "Message of the in-progress UI while saving a Product as draft remotely")
 
+        static let productDuplicatingTitle = NSLocalizedString("Duplicating your product...",
+                                                               comment: "Title of the in-progress UI while duplicated a Product remotely")
```
```suggestion
                                                               comment: "Title of the in-progress UI while duplicating a Product remotely")
```

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormRemoteActionUseCase.swift:
```diff
@@ -24,28 +25,80 @@ final class ProductFormRemoteActionUseCase {
         }
     }
 
+    /// Adds a copy of the input product remotely.
```
Super nit: Would it help to add a point that the new remote product will be in `draft` state?

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormRemoteActionUseCase.swift:
```diff
@@ -24,28 +25,80 @@ final class ProductFormRemoteActionUseCase {
+    func duplicateProduct(originalProduct: EditableProductModel,
+                          password: String?,
+                          onCompletion: @escaping DuplicateProductCompletion) {
+        let productModelToSave: EditableProductModel = {
```
Nit:
```suggestion
        let productModelToSave = {
```

itsmeichigo replied:
I think it's usually good to declare type explicitly, especially for complicated computed variables to avoid long compiling time for type inference. I removed this and it seems that CI (Xcode 13) [fails](https://buildkite.com/automattic/woocommerce-ios/builds/7970#018353f5-ca4e-4823-8d7a-6673769d1a4c) to infer the type, so I restored it back 😅.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormRemoteActionUseCase.swift:
```diff
@@ -205,4 +258,89 @@ private extension ProductFormRemoteActionUseCase {
+                }
+                let newVariation = CreateProductVariation(regularPrice: variation.regularPrice ?? "", attributes: variation.attributes)
+                await self.duplicateProductVariation(newVariation, parent: newProduct)
+            }
```
I think we can send these requests in parallel by using a `TaskGroup` instead of waiting for each variation to finish before duplicating the next variation. 

We can do something like the following,
```suggestion
            await withTaskGroup(of: Void.self) { group in
                for id in variationIDs {
                    group.addTask {
                        guard let variation = await self.retrieveProductVariation(id: id, siteID: newProduct.siteID, productID: oldProductID) else {
                            return
                        }
                        let newVariation = CreateProductVariation(regularPrice: variation.regularPrice ?? "", attributes: variation.attributes)
                        await self.duplicateProductVariation(newVariation, parent: newProduct)
                    }
                }
            }
```

What do you think?

itsmeichigo replied:
Great thinking! 💯  Duplicating variations concurrently makes the process faster - I updated the code following this in 6520428c752093953ad4dabcf49528fc7d161cfe.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormRemoteActionUseCase.swift:
```diff
@@ -205,4 +258,89 @@ private extension ProductFormRemoteActionUseCase {
+                await self.duplicateProductVariation(newVariation, parent: newProduct)
+            }
+
+            let updatedProduct: EditableProductModel = await {
```
Nit:
```suggestion
            let updatedProduct = await {
```

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormRemoteActionUseCase.swift:
```diff
@@ -205,4 +258,89 @@ private extension ProductFormRemoteActionUseCase {
+            DispatchQueue.main.async { [weak self] in
+                self?.stores.dispatch(action)
+            }
+        } as Product?
```
❓ Do we have to explicitly typecast the return values like this? FYI, It builds and runs fine without this in Xcode 14.0. 🤔

itsmeichigo replied:
Yes, it compiles for me too, but since our CI is still running with Xcode 13 I had to add these to make it compiles. There are a few other places in the app that we're doing this - we'll need to remove them when the CI is upgraded to Xcode 14.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormRemoteActionUseCase.swift:
```diff
@@ -205,4 +258,89 @@ private extension ProductFormRemoteActionUseCase {
+
+    func retrieveProduct(id: Int64, siteID: Int64) async -> Product? {
+        await withCheckedContinuation { [weak self] continuation in
+            guard let self = self else { return }
```
Nit: I think we can skip capturing self here as we are capturing it again inside the `DispatchQueue.main.async` call down below.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormRemoteActionUseCase.swift:
```diff
@@ -205,4 +258,89 @@ private extension ProductFormRemoteActionUseCase {
+        } as Product?
+    }
+
+    func retrieveProductVariation(id: Int64, siteID: Int64, productID: Int64) async -> ProductVariation? {
```
Super nit: What do you think about using `variationId` instead of `id` for added clarity? 

```suggestion
    func retrieveProductVariation(variationId: Int64, siteID: Int64, productID: Int64) async -> ProductVariation? {
```

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift:
```diff
@@ -774,6 +781,25 @@ private extension ProductFormViewController {
+            case .success:
+                // Dismisses the in-progress UI, then presents the confirmation alert.
+                self?.navigationController?.dismiss(animated: true, completion: nil)
+                self?.presentProductConfirmationSaveAlert(hasDuplicated: true)
```
❓  Should we present the alert in the completion block as the comment explains? 
```suggestion
                self?.navigationController?.dismiss(animated: true) {
                    self?.presentProductConfirmationSaveAlert(hasDuplicated: true)
                }
```

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/ProductFormViewModelTests.swift:
```diff
@@ -171,6 +171,43 @@ final class ProductFormViewModelTests: XCTestCase {
         XCTAssertFalse(canDeleteProduct)
     }
 
+    // MARK: `canDuplicateProduct`
```
Nice tests. ❤️

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormRemoteActionUseCase.swift:
```diff
@@ -205,4 +258,89 @@ private extension ProductFormRemoteActionUseCase {
+        }
+    }
+
+    func retrieveProduct(id: Int64, siteID: Int64) async -> Product? {
```
❓ I am curious about returning optional here. Do you think throwing instead of returning an optional complicates the code at the calling place?

itsmeichigo replied:
The product is already duplicated before we proceed to duplicate variations, so I'm not sure whether to make this fail if there is a problem in retrieving variations or the updated product. 

But this retrieval is needed to update the UI with the updated product, so maybe it makes sense to throw an error when the data is not updated.

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/ProductFormRemoteActionUseCaseTests.swift:
```diff
@@ -290,6 +290,145 @@ final class ProductFormRemoteActionUseCaseTests: XCTestCase {
     }
+
+    // MARK: - Duplicate a product (`duplicateProduct`)
+    func test_duplicating_product_triggers_adding_copy_of_product_correctly() {
```
Nit: Should we update the test name as we are checking for product status and SKU along with the product name? Or perhaps move those checks to new tests? 😄

itsmeichigo replied:
I think the name is concise enough to indicate that this test checks for the triggered values and not the result of the action.

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/ProductFormRemoteActionUseCaseTests.swift:
```diff
@@ -290,6 +290,145 @@ final class ProductFormRemoteActionUseCaseTests: XCTestCase {
         XCTAssertEqual(result, .failure(.unexpected))
     }
+
+    // MARK: - Duplicate a product (`duplicateProduct`)
```
Nice tests. 👍

selanthiraiyan reviewed (commented):
Thanks for working on this, Huong! 
It is an amazing amount of work in a short period. 🙇 

### Question
- When we duplicate a product, the new draft product still uses the images from the original product. The images are not duplicated. I hope it is the expected behaviour. Could you please confirm? Thanks.

### Code
The code looks great. I left a few comments and questions.

### Testing
I think the failure case isn't working as expected. Please correct me if I am wrong.

The following scenarios happened for a variable product with around 8 variations. 

**Scenario 1**
- I turned off the internet even before tapping the `Duplicate` button, and I ended up on the following screen. 

<img src="https://user-images.githubusercontent.com/524475/190625443-ea8e3115-6592-4e5d-8a9a-8fe6ac586dc2.png" width="300">

- Please notice that the error message says, "Cannot update product"
- Also, I couldn't find the `duplicate_product_failed` event in the Xcode log. 

**Scenario 2**
- I turned off the internet after tapping the `Duplicate` button, and I ended up in a success message saying the product was copied.

<img src="https://user-images.githubusercontent.com/524475/190625641-43bad314-08d5-4db2-b130-cb88c372a258.png" width="300">

- I checked the duplicated product from the `wp-admin` page and found none of the variations was created in the duplicated product. But, there was no error message about the failure.
- Also, I couldn't find the `duplicate_product_failed` event in the Xcode log.

itsmeichigo committed:
Fix incorrect event name when adding product remotely fails

itsmeichigo committed:
Update title for the error alert when duplicating a product fails

itsmeichigo committed:
Send title to product confirmation save alert helper

itsmeichigo committed:
Remove explicit type declaration

itsmeichigo committed:
Rename parameter for retrieveProductVariation

itsmeichigo committed:
Remove redundant self capturing

itsmeichigo committed:
Duplicate variations concurrently

itsmeichigo committed:
Show success alert after dismissing the in-progress UI

itsmeichigo committed:
Make product retrieval failable

itsmeichigo committed:
Restore explicit type declaration to avoid disambiguation

itsmeichigo commented:
Thank you @selanthiraiyan for the detailed reviews!

There was indeed an incorrect tracking for the failure case, and I fixed it in a6b2c39bb426961f142f54ebe36fea6713dbe656 😅.

I can't seem to reproduce the case where the success toast is displayed while the product is not copied. When I turn off the connection after tapping Duplicate, I ended up with the error alert and the copied product is added remotely already.

I addressed the other issues in the comments - this PR is ready for another look!

selanthiraiyan reviewed (approved):
This works great now. 🎉  :shipit:
